services/speech_to_text/v1.js

```diff
@@ -17,54 +17,196 @@
+  var username, password, sUsername, sPassword;
 
   if (service) {
-    username = service.username;
-    password = service.password;
+    sUsername = service.username;
+    sPassword = service.password;
   }
 
+
+  // temp is being used for file streaming to allow the file to arrive so it can be processed. 
+  temp.track();
+
+
+  // These are APIs that the node has created to allow it to dynamically fetch Bluemix
+  // credentials, and also translation models. This allows the node to keep up to 
+  // date with new tranlations, without the need for a code update of this node. 
+
+  // Node RED Admin - fetch and set vcap services
   RED.httpAdmin.get('/watson-speech-to-text/vcap', function (req, res) {
     res.json(service ? {bound_service: true} : null);
   });
 
+
+   // API used by widget to fetch available models
+  RED.httpAdmin.get('/watson-speech-to-text/models', function (req, res) {
+    var stt = watson.speech_to_text({
+      username: username ? username : req.query.un,   
+      password: password ? password : req.query.pwd, 
+      version: 'v1',
+      url: 'https://stream.watsonplatform.net/speech-to-text/api'
+    });
+
+    stt.getModels({}, function(err, models){
+      if (err) {
+        res.json(err);
+      } else {
+        res.json(models);
+      }
+    });          
+  });
+
+  // Utility function to perform a URL validation check
+  function urlCheck(str) {
+    var parsed = url.parse(str)
+
+    return (!!parsed.hostname && !!parsed.protocol && str.indexOf(' ') < 0);
+  }
+
+  // Function that is syncing up the asynchronous nature of the stream
+  // so that the full file can be sent to the API. 
+  var stream_buffer = function(file, contents, cb) {
+    fs.writeFile(file, contents, function (err) {
+      if (err) throw err;
+      cb(fileType(contents).ext);
+    });
+  };
+
+
+  // Function that is syncing up the asynchronous nature of the stream
+  // so that the full file can be sent to the API. 
+  var stream_url = function(file, url, cb) {
+    var wstream = fs.createWriteStream(file);
+
+    wstream.on('finish', function () {
+      fs.readFile(file, function (err, buf) {
+        if (err) {
+          throw err;
+        }
+        cb(fileType(buf).ext)
+      });
+    });
+
+    request(url).pipe(wstream);
+  };
+
+  // This is the Speech to Text Node
+
   function Node (config) {
     RED.nodes.createNode(this, config);
     var node = this;
 
     this.on('input', function (msg) {
-      if (!msg.payload) {
-        var message_err_payload = 'Missing property: msg.payload';
 
-        node.error(message_err_payload, msg)
-        return;
+      // This section is for var functions that will be called, in context with
+      // msg, when the input stream has been received.  
+
+      // This is the callback after the call to the speech to text service.    
+      // Set up as a var within this scope, so it has access to node, msg etc.       
+      var actionComplete = function(err, data) {
```

Comment:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Function 'anonymous' has a complexity of 5.](https://www.codacy.com/app/rezgui-y/node-red-node-watson/file/2636145243/issues/source?bid=3225115&fileBranchId=3319654#l125)